src/doc/tock_binary_format.md

```diff
@@ -586,6 +587,37 @@ Object has both, the kernel decides which to use. Tock is transitioning to
 having the Program Header as the standard one to use, but older kernels (2.0 and
 earlier) do not recognize it and use the Main Header.
 
+#### `10` ShortID
+
+```rust
+struct TbfHeaderV2ShortId {
+    base: TbfHeaderTlv,
+    short_id: u32,
+}
+```
+
+This header allows the compile-time workflow to specify the fixed `ShortId` the
+application should be assigned when it is loaded. The header only includes
+the 32 bit ShortId.
+
+Note, fixed `ShortId`s are defined to be nonzero. Therefore, including this
+header but setting the `short_id` field to 0 is the same as not including this
+header.
+
+Also, this header is just one method for indicating what `ShortId` an
+application should be assigned. The kernel must be configured to use this header
+and any particular Tock kernel may ignore this header.
```

Comment:
This feels confusing to users. Maybe it should be an error as well?

Reply:
I don't know how this could generate an error as it is up to the implementation of `Compress` on how to assign `ShortId`s.

Reply:
Chaining together two recent things, I think the ultimate issue becomes "I as an app have a short ID I expect for myself, and when I use storage, my short ID is used to indicate I wrote/modified things", but now there is a case where the actual short ID written to flash is not the short ID an app expects, which is of particular importance if a _different_ app is reading storage and expecting to see the known identifier and make some decisions based on that.

I suspect there's really two policies that would make sense here, something like `struct TbfHeaderV2ShortIdRequest` and `struct TbfHeaderV2ShortIdMustBe`, where the app can indicate in some way when it has firm requirements in its business logic that it be assigned a specific short ID. (Could also be a third `policy` field in the header added here instead of two headers).

I do, however, ultimately agree that if we are going to use short IDs as the identifiers in storage, we have to be able to give apps a way to (optionally at least) control which short ID they're assigned. Otherwise, there's no way to reliably support cross-app business logic w.r.t. storage.

Reply:
This header is just a convenience feature. The `AppIdPolicy` is the only authority that can decide what AppIDs are assigned. Applications cannot demand specific AppIds (and the permissions those might grant) any more than a linux process can decide it has root permissions.

Reply:
> I don't know how this could generate an error as it is up to the implementation of `Compress` on how to assign `ShortId`s.

If as user compiles an app requesting a specific `shortID`, but then the kernel ignores that and provides a different ID that could cause all sorts of strange issues. Especially if we use `shortID`s for storage. If this doesn't match what the kernel will provide we should error here as something is invalid in the setup. A person went to the effort to specify the `shortID` to use, but the kernel now isn't using it. That could result in incorrect permissions/policies applied later

Reply:
> I suspect there's really two policies that would make sense here, something like `struct TbfHeaderV2ShortIdRequest` and `struct TbfHeaderV2ShortIdMustBe`, where the app can indicate in some way when it has firm requirements in its business logic that it be assigned a specific short ID. (Could also be a third `policy` field in the header added here instead of two headers).
> 
> I do, however, ultimately agree that if we are going to use short IDs as the identifiers in storage, we have to be able to give apps a way to (optionally at least) control which short ID they're assigned. Otherwise, there's no way to reliably support cross-app business logic w.r.t. storage.

I agree. At which point why not just use a `writeID` to avoid having two `shortID`s

Reply:
> This header is just a convenience feature. The `AppIdPolicy` is the only authority that can decide what AppIDs are assigned. Applications cannot demand specific AppIds (and the permissions those might grant) any more than a linux process can decide it has root permissions.

A Linux process can try and use root permissions, it results in an error

Reply:
Is there a proposal to change this text? I think it is completely factual.

Reply:
I think we should be clear that this is invalid and hence an error (the process doesn't load)

Reply:
I think that's imposing a particular application policy and imposing more restrictive semantics on the header. I don't think that's a good idea. How about there is a standard system call (a memop? or a command?) that lets you fetch your ShortId. An application can when it first boots check that its ShortId is what's expect. Would that work?

Reply:
That would help somewhat. But I suspect users will still get confused when they pass `--short-id 0x01` to `elf2tab` but the kernel assigns them a different ID. They will potentially loose data if the shortID changes without their knowledge. It just seems very brittle and confusing